File: test/solo_calendario.js
/**
 * Calendario.gs, la versione solo calendario di Orari.gs, come la prepara Campanella
 *
 *   const { soloCalendario, calendarioDiOggi } = require('./solo_calendario');
 *
 * Chi mette l'orario nel Google Calendar di un altro account (il suo
 * personale, per esempio) incolla in un progetto di quell'account
 * Calendario.gs e un DatiOrari.gs con il solo suo orario. Calendario.gs lo
 * genera l'applicazione da Orari.gs (src/SoloCalendario.cs, SoloCalendario.Genera);
 * qui la stessa regola in JavaScript, per le prove che girano senza
 * l'eseguibile (invarianti_script.js, nomi_funzioni.js,
 * prova_solo_calendario.js). test/prova_orario.ps1 controlla che il file
 * dell'applicazione sia questo, carattere per carattere.
 *
 *   - l'intestazione di Orari.gs (il primo commento, da "/**" in cima al
 *     primo "*" + "/") lascia il posto a src/risorse/Calendario_intestazione.txt,
 *     con {versione} al posto della versione di Orari.gs (_ORARI_VERSIONE);
 *   - le righe fra "// [SOLO EMAIL]" e "// [FINE SOLO EMAIL]" (i segni
 *     compresi) si tolgono: servono solo alle email;
 *   - le righe fra "// [SOLO CALENDARIO.GS]" e "// [FINE SOLO CALENDARIO.GS]"
 *     in Orari.gs sono commenti: perdono il "// " in testa e diventano
 *     codice (i segni si tolgono);
 *   - un segno e' una riga che comincia cosi' (dopo gli spazi); piu' di due
 *     righe vuote di fila diventano due; gli a capo sono \n.
 * Un segno aperto e non chiuso, uno che si chiude senza essere aperto, uno
 * dentro l'altro, una riga "solo calendario" senza "// " o Orari.gs senza la
 * sua intestazione sono errori: l'applicazione non da' un Calendario.gs a meta'.
 */

'use strict';
const fs = require('fs');
const path = require('path');

const SEGNI = {
  email: '// [SOLO EMAIL]',
  fineEmail: '// [FINE SOLO EMAIL]',
  calendario: '// [SOLO CALENDARIO.GS]',
  fineCalendario: '// [FINE SOLO CALENDARIO.GS]'
};

/** Calendario.gs da Orari.gs e dall'intestazione (Calendario_intestazione.txt). */
function soloCalendario(orari, intestazione) {
  const testo = String(orari).replace(/\r\n/g, '\n');
  const v = /\bvar\s+_ORARI_VERSIONE\s*=\s*'([0-9]+(?:\.[0-9]+)*)'/.exec(testo);
  if (!v) throw new Error('in Orari.gs non trovo la versione (_ORARI_VERSIONE)');
  if (testo.indexOf('/**') !== 0) throw new Error('Orari.gs non comincia con la sua intestazione (/**)');
  const fineIntestazione = testo.indexOf('*/');
  if (fineIntestazione < 0) throw new Error('l\'intestazione di Orari.gs non si chiude');
  let resto = testo.slice(fineIntestazione + 2);
  if (resto.charAt(0) === '\n') resto = resto.slice(1);

  const fuori = [];
  let dove = '', riga = 0;
  const vuota = s => s.replace(/[ \t]+/g, '') === '';
  for (const r of resto.split('\n')) {
    riga++;
    const t = r.replace(/^[ \t]+/, '');
    const segno = (t.indexOf(SEGNI.email) === 0) ? 'email'
                : (t.indexOf(SEGNI.fineEmail) === 0) ? 'fineEmail'
                : (t.indexOf(SEGNI.calendario) === 0) ? 'calendario'
                : (t.indexOf(SEGNI.fineCalendario) === 0) ? 'fineCalendario' : '';
    if (segno === 'email' || segno === 'calendario') {
      if (dove) throw new Error('riga ' + riga + ' dopo l\'intestazione: ' + SEGNI[segno] + ' dentro un altro blocco');
      dove = segno;
      continue;
    }
    if (segno === 'fineEmail' || segno === 'fineCalendario') {
      if (dove !== (segno === 'fineEmail' ? 'email' : 'calendario')) {
        throw new Error('riga ' + riga + ' dopo l\'intestazione: ' + SEGNI[segno] + ' senza il suo inizio');
      }
      dove = '';
      continue;
    }
    if (dove === 'email') continue;
    let scritta = r;
    if (dove === 'calendario') {
      const rientro = r.slice(0, r.length - t.length);
      if (t === '//') scritta = '';
      else if (t.indexOf('// ') === 0) scritta = rientro + t.slice(3);
      else throw new Error('riga ' + riga + ' dopo l\'intestazione: nel blocco solo calendario manca "// " in testa');
    }
    // piu' di due righe vuote di fila diventano due
    if (vuota(scritta) && fuori.length >= 2 && vuota(fuori[fuori.length - 1]) && vuota(fuori[fuori.length - 2])) continue;
    fuori.push(scritta);
  }
  if (dove) throw new Error('il blocco ' + SEGNI[dove] + ' non si chiude');

  const testa = String(intestazione).replace(/\r\n/g, '\n').split('{versione}').join(v[1]).replace(/\s+$/, '');
  return testa + '\n' + fuori.join('\n');
}

/** Calendario.gs dai file di adesso: src/risorse/Orari.gs e Calendario_intestazione.txt. */
function calendarioDiOggi() {
  const risorse = path.join(__dirname, '..', 'src', 'risorse');
  return soloCalendario(fs.readFileSync(path.join(risorse, 'Orari.gs'), 'utf8'),
                        fs.readFileSync(path.join(risorse, 'Calendario_intestazione.txt'), 'utf8'));
}

module.exports = { SEGNI, soloCalendario, calendarioDiOggi };

File: src/SoloCalendario.cs
// ===========================================================================
//  SoloCalendario.cs - l'orario nel Google Calendar di un altro account
//
//  Lo script della Posta e degli Orari gira nell'account della scuola. Chi
//  usa il Google Calendar di un altro account (il suo personale, per
//  esempio) mette l'orario da li': in un progetto Apps Script di
//  quell'account incolla Calendario.gs, la versione solo calendario di
//  Orari.gs, e un DatiOrari.gs con il solo suo orario
//  (AnalisiOrario.GeneraDatiDelDocenteGs). Calendario.gs nasce da Orari.gs,
//  la risorsa incorporata:
//
//    - l'intestazione di Orari.gs (il primo commento, da "/**" al primo
//      "*/") lascia il posto a Calendario_intestazione.txt, con la versione
//      di Orari.gs al posto di {versione};
//    - le righe fra "// [SOLO EMAIL]" e "// [FINE SOLO EMAIL]" (i segni
//      compresi) si tolgono: servono solo alle email (l'invio, i messaggi,
//      l'etichetta, il tuo indirizzo);
//    - le righe fra "// [SOLO CALENDARIO.GS]" e "// [FINE SOLO CALENDARIO.GS]"
//      in Orari.gs sono commenti: qui perdono il "// " in testa e diventano
//      codice;
//    - un segno e' una riga che comincia cosi' (dopo gli spazi); piu' di due
//      righe vuote di fila diventano due; gli a capo sono \n.
//
//  test\solo_calendario.js fa la stessa cosa in JavaScript, per le prove
//  che girano senza l'eseguibile, e test\prova_orario.ps1 controlla che le
//  due diano lo stesso file. Se Calendario.gs nominasse ancora un servizio
//  delle email o dell'indirizzo (o uno che Campanella non usa, come il
//  Drive), Google chiederebbe anche quel permesso nell'altro account: Genera
//  allora si ferma, e l'applicazione non lo da'.
// ===========================================================================

using System;
using System.Collections.Generic;
using System.IO;
using System.Text.RegularExpressions;

namespace Campanella
{
    static class SoloCalendario
    {
        public const string NomeFile = "Calendario.gs";
        public const string RisorsaIntestazione = "Calendario_intestazione.txt";

        public const string SegnoEmail = "// [SOLO EMAIL]";
        public const string FineEmail = "// [FINE SOLO EMAIL]";
        public const string SegnoCalendario = "// [SOLO CALENDARIO.GS]";
        public const string FineCalendario = "// [FINE SOLO CALENDARIO.GS]";

        /// <summary>
        /// I nomi che in Calendario.gs non ci devono essere, nemmeno in un
        /// commento: i servizi delle email e dell'indirizzo, e quelli che
        /// nessuno script di Campanella usa.
        /// </summary>
        public static readonly string[] Vietati =
        {
            "MailApp", "GmailApp", "Gmail", "sendEmail", "getActiveUser", "getEffectiveUser",
            "UrlFetchApp", "DriveApp", "DocumentApp", "SpreadsheetApp", "FormApp"
        };

        /// <summary>
        /// Calendario.gs dagli script incorporati. Se non si puo' preparare, un
        /// commento che dice perche' (incollato, non fa niente).
        /// </summary>
        public static string Codice()
        {
            string orari = Guscio.LeggiRisorsa("Orari.gs");
            string testa = Guscio.LeggiRisorsa(RisorsaIntestazione);
            if (orari == "" || testa == "") return "// Risorsa non trovata: ricompila l'applicazione.";
            try { return Genera(orari, testa); }
            catch (InvalidDataException ex)
            {
                return "// Non riesco a preparare Calendario.gs: " + ex.Message + ".\n" +
                       "// Per adesso metti l'orario dall'account della scuola, con Orari.gs.";
            }
        }

        /// <summary>
        /// Calendario.gs da Orari.gs e dall'intestazione: vedi in cima al file.
        /// Un segno aperto e non chiuso, uno che si chiude senza essere aperto,
        /// uno dentro l'altro, una riga "solo calendario" senza "// ", Orari.gs
        /// senza intestazione o senza versione, un nome vietato nel risultato:
        /// InvalidDataException, con il motivo.
        /// </summary>
        public static string Genera(string orari, string intestazione)
        {
            string testo = (orari ?? "").Replace("\r\n", "\n");
            Match v = Regex.Match(testo, @"\bvar\s+_ORARI_VERSIONE\s*=\s*'([0-9]+(?:\.[0-9]+)*)'");
            if (!v.Success) throw new InvalidDataException("in Orari.gs non trovo la versione (_ORARI_VERSIONE)");
            if (!testo.StartsWith("/**", StringComparison.Ordinal))
                throw new InvalidDataException("Orari.gs non comincia con la sua intestazione (/**)");
            int fineIntestazione = testo.IndexOf("*/", StringComparison.Ordinal);
            if (fineIntestazione < 0) throw new InvalidDataException("l'intestazione di Orari.gs non si chiude");
            string resto = testo.Substring(fineIntestazione + 2);
            if (resto.StartsWith("\n", StringComparison.Ordinal)) resto = resto.Substring(1);

            List<string> fuori = new List<string>();
            string dove = "";
            int riga = 0;
            foreach (string r in resto.Split('\n'))
            {
                riga++;
                string t = r.TrimStart(' ', '\t');
                string segno = t.StartsWith(SegnoEmail, StringComparison.Ordinal) ? "email"
                             : t.StartsWith(FineEmail, StringComparison.Ordinal) ? "fineEmail"
                             : t.StartsWith(SegnoCalendario, StringComparison.Ordinal) ? "calendario"
                             : t.StartsWith(FineCalendario, StringComparison.Ordinal) ? "fineCalendario" : "";
                if (segno == "email" || segno == "calendario")
                {
                    if (dove != "")
                        throw new InvalidDataException("riga " + riga + " dopo l'intestazione: " + Segno(segno) +
                                                       " dentro un altro blocco");
                    dove = segno;
                    continue;
                }
                if (segno == "fineEmail" || segno == "fineCalendario")
                {
                    if (dove != (segno == "fineEmail" ? "email" : "calendario"))
                        throw new InvalidDataException("riga " + riga + " dopo l'intestazione: " + Segno(segno) +
                                                       " senza il suo inizio");
                    dove = "";
                    continue;
                }
                if (dove == "email") continue;
                string scritta = r;
                if (dove == "calendario")
                {
                    string rientro = r.Substring(0, r.Length - t.Length);
                    if (t == "//") scritta = "";
                    else if (t.StartsWith("// ", StringComparison.Ordinal)) scritta = rientro + t.Substring(3);
                    else throw new InvalidDataException("riga " + riga + " dopo l'intestazione: nel blocco solo " +
                                                        "calendario manca \"// \" in testa");
                }
                // piu' di due righe vuote di fila diventano due
                int n = fuori.Count;
                if (Vuota(scritta) && n >= 2 && Vuota(fuori[n - 1]) && Vuota(fuori[n - 2])) continue;
                fuori.Add(scritta);
            }
            if (dove != "") throw new InvalidDataException("il blocco " + Segno(dove) + " non si chiude");

            string testa = (intestazione ?? "").Replace("\r\n", "\n").Replace("{versione}", v.Groups[1].Value).TrimEnd();
            string codice = testa + "\n" + string.Join("\n", fuori.ToArray());
            foreach (string nome in Vietati)
            {
                Match m = Regex.Match(codice, @"(?<![\w$])" + Regex.Escape(nome) + @"(?![\w$])");
                if (m.Success)
                    throw new InvalidDataException("nominerebbe " + nome + " (riga " +
                        (codice.Substring(0, m.Index).Split('\n').Length) + "), e Google ne chiederebbe il permesso");
            }
            return codice;
        }

        static string Segno(string quale)
        {
            switch (quale)
            {
                case "email": return SegnoEmail;
                case "fineEmail": return FineEmail;
                case "calendario": return SegnoCalendario;
                default: return FineCalendario;
            }
        }

        static bool Vuota(string s) { return s.Replace(" ", "").Replace("\t", "") == ""; }
    }
}
